castai/resource_commitments_mapping.go

```diff
@@ -0,0 +1,416 @@
+package castai
+
+import (
+	"errors"
+	"fmt"
+	"path"
+	"strconv"
+	"time"
+
+	"github.com/samber/lo"
+	"golang.org/x/exp/slices"
+
+	"github.com/castai/terraform-provider-castai/castai/sdk"
+)
+
+type (
+	// Terraform SDK's diff setter uses mapstructure under the hood
+
+	gcpCUDResource struct {
+		// CAST AI only fields
+		ID             *string  `mapstructure:"id,omitempty"` // ID of the commitment
+		AllowedUsage   *float32 `mapstructure:"allowed_usage,omitempty"`
+		Prioritization *bool    `mapstructure:"prioritization,omitempty"`
+		Status         *string  `mapstructure:"status,omitempty"`
+
+		// Fields from GCP CUDs export JSON
+		CUDID          string `mapstructure:"cud_id"` // ID of the CUD in GCP
+		CUDStatus      string `mapstructure:"cud_status"`
+		StartTimestamp string `mapstructure:"start_timestamp"`
+		EndTimestamp   string `mapstructure:"end_timestamp"`
+		Name           string `mapstructure:"name"`
+		Region         string `mapstructure:"region"`
+		CPU            int    `mapstructure:"cpu"`
+		MemoryMb       int    `mapstructure:"memory_mb"`
+		Plan           string `mapstructure:"plan"`
+		Type           string `mapstructure:"type"`
+	}
+	gcpCUDConfigResource struct {
+		Matcher        []*gcpCUDConfigMatcherResource `mapstructure:"matcher,omitempty"`
+		Prioritization *bool                          `mapstructure:"prioritization,omitempty"`
+		Status         *string                        `mapstructure:"status,omitempty"`
+		AllowedUsage   *float32                       `mapstructure:"allowed_usage,omitempty"`
+	}
+	gcpCUDConfigMatcherResource struct {
+		Name   string  `mapstructure:"name"`
+		Type   *string `mapstructure:"type,omitempty"`
+		Region *string `mapstructure:"region,omitempty"`
+	}
+
+	azureReservationResource struct {
+		ID            *string `mapstructure:"id,omitempty"`   // ID of the commitment
+		ReservationID string  `mapstructure:"reservation_id"` // ID of the reservation in Azure
+	}
+
+	// commitmentResource is an interface for common management of GCP (gcpCUDResource) and Azure (azureReservationResource) resources
+	commitmentResource interface {
+		// GetCommitmentID returns the ID of the commitment in CAST AI
+		GetCommitmentID() string
+		// GetIDInCloud returns the ID of the resource in the cloud provider
+		GetIDInCloud() string
+	}
+)
+
+func (r *gcpCUDConfigResource) GetMatcher() *gcpCUDConfigMatcherResource {
```

Comment:
Could these functions stay private as well? Unless it is to satisfy some Go interface, which I think is only the case for two funcs of `commitmentResource`, right?.